### PR TITLE
🎨 `print` statements → `log` statements

### PR DESCRIPTION
**Pull request by shnizzedy**
## Fixes

Related to https://github.com/FCP-INDI/C-PAC/pull/2048 by @nx10
Related to https://github.com/FCP-INDI/C-PAC/pull/2058 by @shnizzedy

## Description

> `print` statements are useful in some situations (e.g., debugging), but should typically be omitted from production code. `print` statements can lead to the accidental inclusion of sensitive information in logs, and are not configurable by clients, unlike `logging` statements.

― https://docs.astral.sh/ruff/rules/print/

> Like `print` statements, `pprint` statements are useful in some situations (e.g., debugging), but should typically be omitted from production code. `pprint` statements can lead to the accidental inclusion of sensitive information in logs, and are not configurable by clients, unlike `logging` statements.

― https://docs.astral.sh/ruff/rules/p-print/

## Technical details

* Because this PR touches so many files, I'm mostly not taking the time to otherwise lint them as part of this change.
* https://github.com/FCP-INDI/C-PAC/blob/71c9e6fa7bc82d618088a6aaa129b812257ca3a6/.ruff.toml#L11 standardizes `import nibabel as nib` in 71c9e6fa7bc82d618088a6aaa129b812257ca3a6 (in some places it was `import nibabel as nb`), but ruff doesn't catch the nipype-style import strings like https://github.com/FCP-INDI/C-PAC/blob/71c9e6fa7bc82d618088a6aaa129b812257ca3a6/CPAC/distortion_correction/distortion_correction.py#L500; those are also fixed in this PR
* For Function nodes, automatically inserts this line https://github.com/FCP-INDI/C-PAC/blob/e460f7d661613703a06f80e4087cedb0104fae14/CPAC/utils/interfaces/function/function.py#L65-L66 after the imports so `FMLOGGER`, `IFLOGGER`, `UTLOGGER`, and `WFLOGGER` are always available in Function nodes as the `"nipype.filemanip"`, `"nipype.interface"`, `"nipype.utils"`, and `"nipype.workflow"` loggers, respectively

## Tests

https://github.com/FCP-INDI/C-PAC/blob/e460f7d661613703a06f80e4087cedb0104fae14/CPAC/utils/interfaces/tests/test_function.py#L23-L50

## Checklist

- [x] My pull request has a descriptive title (not a vague title like `Update index.md`).
- [x] My pull request targets the `lint/pep8-boolean-comparison` branch of the repository. 
- [x] My commit messages follow best practices.
- [x] My code follows the established code style of the repository.
- [x] I added tests for the changes I made.
- [ ] I updated the changelog.
- [ ] I added or updated documentation (if applicable).
- [ ] I tried running the project locally and verified that there are no visible errors.

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

File: CPAC/utils/ndmg_utils.py [hunk 1]
```diff
@@ -183,8 +189,9 @@ def cor_graph(self, timeseries, attr=None):
         """
         import numpy as np
 
-        timeseries[0]
+        ts = timeseries[0]  # noqa: F841
```
Comment (shnizzedy) [inline CPAC/utils/ndmg_utils.py:192]: This `ts` isn't used anywhere. Do we care to assign it just to annotate, or should we drop this line?

**Review by shnizzedy — COMMENTED** [review]
Where I came upon the pattern:

```Python
print(error_message)
sys.exit(exitcode)
```

, I wasn't consistent. Mostly I updated to something like:

```Python
logger.error(error_message)
sys.exit(exitcode)
```

but sometimes I did like:

```Python
raise Exception(error_message)
```

(but with a more specific error class than `Exception`). I think we generally want to convert all the `sys.exit`s to `raise Exception`s, but this PR's already uncomfortably large without doing so. I think that can be a subsequent PR if we care to.

---

I mostly used the `"nipype.workflow"` logger here, unless a module was already loading `"nipype.interface"` or the logger was just for a standalone test.

We might want to have more of these logs logged in `"nipype.interface"` or another logger, and we might want to make those changes before we merge this PR?

---

**Comment by shnizzedy** [comment]
Oh, I forgot to mention this in the PR message, but updating the Function nodes reminded me of something I wondered ages ago, then forgot about:

Why don't we default https://github.com/FCP-INDI/C-PAC/blob/26ebd1ebb88b15229577757befda121b8e4490ab/CPAC/utils/interfaces/function/function.py#L163 as `True` (or even not make it an option, and just always have Function nodes "as module"s)?

<span title="That is">I.e.</span>, when would we want a Function node not be "as module"?

File: CPAC/utils/ndmg_utils.py (hunk 1 above)
Comment (sgiavasis) [reply CPAC/utils/ndmg_utils.py:192]: It's fine for now. Your comments with idea stubs in this PR is a pretty nice list, so we can revisit this when we revisit the other ideas as well.

**Comment by sgiavasis** [comment]
> Oh, I forgot to mention this in the PR message, but updating the Function nodes reminded me of something I wondered ages ago, then forgot about:
> 
> Why don't we default
> 
> https://github.com/FCP-INDI/C-PAC/blob/26ebd1ebb88b15229577757befda121b8e4490ab/CPAC/utils/interfaces/function/function.py#L163
> 
> as `True` (or even not make it an option, and just always have Function nodes "as module"s)?
> I.e., when would we want a Function node not be "as module"?

This is ancient; I don't remember right now but again, can add to the list of things to revisit in this PR.

**Comment by shnizzedy** [comment]
> [[This](https://github.com/FCP-INDI/C-PAC/pull/2062#issuecomment-1954443008)] is ancient; I don't remember right now but again, can add to the list of things to revisit in this PR.

― https://github.com/FCP-INDI/C-PAC/pull/2062#issuecomment-1970048918

It's cool; as far as I can tell, it
1. Includes the variables available in the namespace like if you imported the function normally outside of nipype (so you don't need the `imports` param or the imports defined in the function itself
2. Includes the module name in the log instead of just the text of the function

But I don't know how much overhead this costs or saves, or what other considerations we should have about whether / when to use it

**Pull request merged**